addons/mail/static/src/components/activity/activity_tests.js

```diff
@@ -1,6 +1,6 @@
 /** @odoo-module **/
 
-import { useModels } from '@mail/component_hooks/use_models/use_models';
+import { registerMessagingComponent } from '@mail/utils/messaging_component';
 import { Activity } from '@mail/components/activity/activity';
```

Comment:
The registry is not exposed to the outside in a way that's easily accessible, and I'm not sure there's a great reason to expose it. Maybe this test shouldn't even use the registry at all and pass the Activity component explicitly to the ParentComponent.

Reply:
Only components that are coming from the registry should be trusted, because this is where they are build/decorated.

Import works now because the registry currently updates the components in place, but I think it's a mistake. It should register a new version of the component instead. It's not critical to change it now, but the rest of the code should not be based on the current implementation.

Reply:
In the case where we don't do the change in place, I think we should export the decorated component from the file that defines it instead of the undecorated one anyway, and this point would not really apply.

I don't think in unit tests we should use whatever is currently in the registry depending on what modules are installed, as this means that tests on features that are overriden by other modules cannot exist. Whether all overrides work together should be handled by integration tests (tours).

Currently overrides have to be in place anyway because of template inheritance being done server side, but there are plans to add xpath processing capabilities to the client which would allow to test non-overriden and overriden components in the same test-suite, which would be more robust IMO.

Reply:
> In the case where we don't do the change in place, I think we should export the decorated component from the file that defines it instead of the undecorated one anyway, and this point would not really apply.

This is a mistake, we should always use the registered version, not the file itself which is just definition.
If you want an analogy with python models, you use them from the registry, not by importing the definition file.
It's pretty much the same for any registry actually.

> I don't think in unit tests we should use whatever is currently in the registry depending on what modules are installed, as this means that tests on features that are overriden by other modules cannot exist. Whether all overrides work together should be handled by integration tests (tours).

I agree, but the proper change to that is to stop registering the components automatically in their own file.
What must be done instead is to keep the component files as just definition, and then register them in the boot file only imported in the prod bundle (actual `main.js`, or it could also be a service instead), and to have a separate registry for the each test.
If this was the chosen architecture, then yes it would be fine to keep importing the component definition here instead of the registry result. But as it's not the chosen architecture, it's not ok.
Note that in the same vein, the "parent component" should definitely not be registered in the main registry then, and we wouldn't need to worry about unregister at all, just discard the test registry once it's done.

> Currently overrides have to be in place anyway because of template inheritance being done server side, but there are plans to add xpath processing capabilities to the client which would allow to test non-overriden and overriden components in the same test-suite, which would be more robust IMO.

I'm looking forward to the change, I've always wanted to be able to test our components with and without overrides applied.

Reply:
So please either use the prod registry as I suggested originally, as a technical limitation for now.
Or go all the way and create a separate test registry (but I don't think it's worth it to delay the PR for this part).

Just avoid using the definition file directly. We want to be able to change the way we define components internally without having to adapt the rest of the code.

Reply:
> I agree, but the proper change to that is to stop registering the components automatically in their own file.
> What must be done instead is to keep the component files as just definition, and then register them in the boot file only imported in the prod bundle (actual `main.js`, or it could also be a service instead), and to have a separate registry for the each test.

The way we do it in web is that we register things in their definition file but empty registries before the tests and register explicitly what we're testing in the test. This preserves locality better than splitting the definition and registration and gives us the same control over the test environment. You can take a look at `prepareRegistriesWithCleanup` if you'd like more details on how this works.

I'm not too particular about the details, just wanted to raise those points so you're aware of it but I'll adapt the implementation as you're suggesting, even if I don't agree entirely on everything, I probably won't be the person who will need to maintain it in the future so it's better if it conforms to your plans and expectations as much as possible.

Reply:
Interesting, if you see a way to do it like that I'm ok for it. I like to keep the locality too if possible, it makes things much easier.

The main point for me is that I want to be able to make the component.js file return something else without impacting anywhere else than in the register function that is made to receive its content.

Reply:
The best way to ensure it works like I would expect is to change the register function to create a copy of the component class to add the messaging features, instead of updating it in place.